graph/path/yen_ksp.go

```diff
@@ -0,0 +1,148 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package path
+
+import (
+	"sort"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/internal/set"
+)
+
+
+// YenKShortestPath returns the k-shortest loopless paths from s to t in g. YenKShortestPath will
+// panic if g contains a negative edge weight.
+func YenKShortestPath(g graph.Graph, k int, s, t graph.Node) [][]graph.Node {
+	yk := yenKSPAdjuster{
+		g:       g,
+		visited: make(map[int64]set.Int64s),
+	}
+
+	if wg, ok := g.(Weighted); ok {
+		yk.weight = wg.Weight
+	} else {
+		yk.weight = UniformCost(g)
+	}
+
+	shortest, _ := DijkstraFrom(s, yk).To(t.ID())
+	if len(shortest) == 0 {
+		return nil
+
+	}
+	paths := [][]graph.Node{shortest}
+
+	var pot []yenShortest
+
+	for i := int64(1); i < int64(k); i++ {
+		for n := 0; n < (len(paths[i-1]) - 1); n++ {
+			spur := paths[i-1][n]
+			root := make([]graph.Node, len(paths[i-1][:n+1]))
+			copy(root, paths[i-1][:n+1])
+
+			var rootWeight float64
+			for x := 1; x < len(root); x++ {
+				w, _ := yk.weight(root[x-1].ID(), root[x].ID())
+				rootWeight += w
+			}
+
+			for _, path := range paths {
+				if len(path) > n {
+					ok := true
+					for x := 0; x < len(root); x++ {
+						if path[x].ID() != root[x].ID() {
+							ok = false
+							break
+						}
+					}
+					if ok {
+						yk.addVisited(path[n].ID(), path[n+1].ID())
+					}
+				}
+			}
+
+			spath, weight := DijkstraFrom(spur, yk).To(t.ID())
+
+			if len(root) > 1 {
+				root = append(root[:len(root)-1], spath...)
+				pot = append(pot, yenShortest{root, weight + rootWeight})
+			} else {
+				pot = append(pot, yenShortest{spath, weight})
+			}
+
+			yk.visited = make(map[int64]set.Int64s)
+		}
+
+		if len(pot) == 0 {
+			break
+		}
+
+		sort.Sort(byPathWeight(pot))
+		paths = append(paths, pot[0].path)
+		pot = pot[1:]
+	}
+
+	return paths
+}
+
+type yenShortest struct {
+	path      []graph.Node
+	weight float64
```

Comment:
Please gofmt your code (can I recommend that you add an on-save hook to goimports or gofmt - then you don't need to think about doing this)